olp-cpp-sdk-dataservice-read/src/StreamLayerClientImpl.cpp

```diff
@@ -155,6 +155,78 @@ client::CancellableFuture<SubscribeResponse> StreamLayerClientImpl::Subscribe(
       std::move(cancel_token), std::move(promise));
 }
 
+client::CancellationToken StreamLayerClientImpl::Unsubscribe(
+    UnsubscribeResponseCallback callback) {
+  auto unsubscribe_task =
+      [=](client::CancellationContext context) -> UnsubscribeResponse {
+    std::string subscription_id;
+    std::string subscription_mode;
+    std::string node_base_url;
+    std::string x_correlation_id;
+
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      if (!client_context_) {
+        OLP_SDK_LOG_WARNING_F(kLogTag,
+                              "Unubscribe: unsuccessful, not subscribed");
+
+        return client::ApiError(client::ErrorCode::PreconditionFailed,
+                                "Not subscribed", false);
+      }
+
+      subscription_id = client_context_->subscription_id;
+      subscription_mode = client_context_->subscription_mode;
+      node_base_url = client_context_->node_base_url;
+      x_correlation_id = client_context_->x_correlation_id;
+    }
+
+    OLP_SDK_LOG_INFO_F(
+        kLogTag,
+        "Unsubscribe: started, subscription_id=%s, subscription_mode=%s, "
+        "node_base_url=%s, x_correlation_id=%s",
+        subscription_id.c_str(), subscription_mode.c_str(),
+        node_base_url.c_str(), x_correlation_id.c_str());
+
+    client::OlpClient client;
+    client.SetBaseUrl(node_base_url);
+    client.SetSettings(settings_);
```

Comment:
It seems to me like a waste of resources to always create this on the fly.
I would propose for the next review lets wrap it into a shared_ptr and create it once during Subscribe().